src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -3684,7 +3684,8 @@ void CallData::ApplyServiceConfigToCallLocked(grpc_call_element* elem) {
     // from the client API, reset the deadline timer.
     if (chand->deadline_checking_enabled() && method_params_->timeout() != 0) {
       const grpc_millis per_method_deadline =
-          grpc_timespec_to_millis_round_up(call_start_time_) +
+          grpc_timespec_to_millis_round_up(
```

Comment:
This takes the int64, converts to double, converts back to int64/32, then sends it to a method which does a bunch more divisions with it as a double.

Would it perhaps make more sense to go directly from cycles to millis, without any converts to/from double?

ie. int64 grpc_millis_from_cycle_clock(int64_t cycle_count);

Reply:
Good point Arjun, and I agree this is no optimal.  That said, unfortunately, the problem is that `grpc_timespec_to_millis*` is based on IOMGR's `g_start_time` as its origin, and `gpr_cycle_counter` is based on clock precise's `start_cycle` or `CLOCK_REAL`. 

We will have to convert `CLOCK_PRECISE` to `CLOCK_MONOTONIC` to calculate the time span and then calculate the absolute deadline in `grpc_millis`.  Since we don't have access to `g_start_time` and we cannot make assumption about its clock type, we had to convert the cycles to time :-/

Since this only impacts services with `ServiceConfig`, do you mind if we deal with it later if it became a bottleneck?

Reply:
My objection at present is that right now, with your change, it seems like the following sequence of events would happen:

call_start_time_ = rdtsc();
<...>
timespec_precise = gpr_cycle_counter_to_time(call_start_time_);

grpc_timespec_to_millis_round_up(timespec_precise)
 => calls gpr_convert_clock_type(from=precise, to=monotonic)
=>   return gpr_time_add(gpr_now(monotonic),
                      gpr_time_sub(t, gpr_now(precise)));

So we're doing a gpr_now(monotonic) and a gpr_now(precise) in this codepath, and also, one that is susceptible to heavy scheduling inaccuracy.

My alternative proposal:

at init time:

double cycles_per_sec = <...>;
double cycles_per_nano = cycles_per_sec / 1E9;
double cycles_per_milli = cycles_per_sec / 1E3;

<...>
clock_a = rdtsc();
g_start_time_ = gpr_now(monotonic);
clock_b = rdtsc();
// Bracket and average. You could even do this a bunch of times to be more confident.
g_start_clock = avg(clock_a, clock_b);

Then we could have:

int64_t cycles_to_wall_msec(int64_t cycle_count) {
  return (cycle_count - g_start_clock) / cycles_per_milli;
}

Reply:
That's a brilliant idea. Thank you for the suggestion! PTAL.

Reply:
Looks fine - optionally we could also consider saving a couple of float divides (ie. not converting to timespec in clock_precise.cc before converting back to an int64 millis in exec_ctx.cc) but I won't block the review on that.

Reply:
Thanks. That would require exposing the frequency, which I have tried to avoid.  If we see this anywhere in profiles, we can optimize it.